### ISSUE TITLE
Using Lerna with Storybook and symlink dependencies

### ISSUE THREAD

**Issue by stevenfitzpatrick** [lerna/lerna#1367]
Hey,

I am trying to figure out how to use Lerna in combination with Storybook where I have React Components in different packages. Some have dependencies on each other.

I know I can run `lerna bootstrap` but this set symlink in the node_modules of the respective packages and not the root.

My structure looks a bit like this

```
- packages
  - Icons
    - src
         Icon.jsx
         Icon.stories.js
    package.json
  - Button
    - src
         Button.jsx
         Button.stories.js
    package.json
```

And imagine my button component has a dependency on icon package. So it will have an import like this

`import { Icon } from '@myscope/icons'`

Keep in mind I have never published my packages before.

And then when I run my storybook story of button it throws an errror like this

`Module not found: Error: Can't resolve '@myscope/icons' in '/ui-component-lib/packages/button/src/button'`

Any idea how I get can Lerna and Storybook to work together using symlinks ?

**Comment by evocateur** [comment]
As long as `@myscope/button` has a dependency on `@myscope/icons`, `lerna bootstrap` should create the correct symlinks (even for unpublished modules) such that Storybook "just works". I'm afraid I'll need more context (`lerna.json`, specific commands that were run, any `lerna-debug.log` output, etc) to accurately diagnose the issue.

**Comment by evocateur** [comment]
(There's nothing special about Storybook, if you can import sibling packages into the button component and run it bundled in a browser, it should work in the tests)

**Comment by stevenfitzpatrick** [comment]
`@myscope/button` does have the correct symlink dependency to `@myscope/icons` and I can see this if I open the node_modules folder within `@myscope/button`.

But my feeling is that storybook only resolves packages within the root `node_modules` not the nested `node_modules` within the packages folders.

So in storybook when I have `@myscope/icons` which I don't have in my root folder (and never been published) it basically goes...cannot find this module.

I seem to have found one workaround in webpack.config.js for storybook is to add an alias to the package like this

```
 resolve: {
    alias: {
      '@myscope/icons': path.resolve(__dirname, '../packages/icons/src/index')
    }
  },
```

**Comment by evocateur** [comment]
That's really weird, such an alias shouldn't be necessary at all (you shouldn't need to webpack at all just to run tests?). I wasn't aware Storybook ignores `node_modules` resolution semantics. I would raise the bug with them.

**Comment by stevenfitzpatrick** [comment]
Well this is my guess trying to get these symlinked lerna packages working in Storybook.

**Comment by stevenfitzpatrick** [comment]
Is there anyway to add the lerna packages as `symlink` to the root node_modules by the way.

`lerna bootstrap` just add the symlink in node_modules of that package.

**Comment by evocateur** [comment]
I use Jest from the root literally every day with lerna bootstrapped local modules without issue. You've got some config somewhere polluting things. There is absolutely no requirement for the symlinks to be in the root `node_modules`.

**Comment by stevenfitzpatrick** [comment]
Ill set up sample repo, see if I can reproduce.

**Comment by stevenfitzpatrick** [comment]
Ah I think I found the problem.

In the package.json of my packages I have this

  "name": "@myscope/icons",
  "main": "dist/index.js",

But at this time I do not have a `dist` folder yet because I haven't run the build command yet. Since I just looking to run `jest` at this time.

So if I change `main` to "src/index.js" it works.

Is there a way to have some logic for this for running local tests and storybook that package.json points to `src/index.js` and when deployed to look for `dist/index.js`

**Comment by evocateur** [comment]
No, you have to transpile before running "integration" tests like that. I usually run it from the root in a `"pretest"` lifecycle.

**Comment by stevenfitzpatrick** [comment]
Oh ok, so this is a pretty normal problem then, and when I build the symlink folder should update and get a `dist` folder yeah ?

Thanks for all your help by the way.

**Comment by evocateur** [comment]
It's a normal situation when compiling the output of an npm module (whether in a lerna monorepo or a "normal" npm package repository). I usually add a `prepare` lifecycle to the packages that transpiles immediately after `lerna bootstrap`, and then still do a recompile in `pretest`.
